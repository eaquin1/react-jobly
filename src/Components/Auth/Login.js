import React, { useState } from "react";

import { Label, Input, Form, Button, FormGroup, Container } from "reactstrap";
import JoblyApi from "../../Helpers/JoblyApi";
import { useHistory } from "react-router-dom";

function Login({setToken}) {
    const history = useHistory();
    const INITIAL_STATE_LOGIN = {
        username: "",
        password: "",
        first_name: "",
        last_name: "",
        email: "",
        errors: [],
    };
    const [loginFormData, setLoginFormData] = useState(INITIAL_STATE_LOGIN);
    const [activeView, setActiveView] = useState("login");

    function setLoginView() {
        setActiveView("login");
    }

    function setRegisterView() {
        setActiveView("register");
    }

    const handleChange = (e) => {
        const { name, value } = e.target;
        setLoginFormData((formData) => ({
            ...formData,
            [name]: value,
        }));
    };

    async function handleLogin(evt) {
        evt.preventDefault();
        let data;
        let endpoint;

        if (activeView === "register") {
            //unrequired fields can be undefined
            data = {
                username: loginFormData.username,
                password: loginFormData.password,
                first_name: loginFormData.first_name || undefined,
                last_name: loginFormData.last_name || undefined,
                email: loginFormData.email || undefined,
            };
            endpoint = "register";
        } else {
            data = {
                username: loginFormData.username,
                password: loginFormData.password,
            };
            endpoint = "login";
        }
        let token;
        try {
            token = await JoblyApi[endpoint](data);
        } catch (errors) {
            return setLoginFormData((l) => ({ ...l, errors }));
        }

        setToken(token)
        history.push("/jobs");
    }

    const registerFields = (
        <>
            <FormGroup>
                <Label for="first_name">First Name</Label>
                <Input
                    type="text"
                    name="first_name"
                    placeholder="First Name"
                    value={loginFormData.first_name}
                    onChange={handleChange}
                />
            </FormGroup>
            <FormGroup>
                <Label for="last_name">Last Name</Label>
                <Input
                    type="text"
                    name="last_name"
                    placeholder="Last Name"
                    value={loginFormData.last_name}
                    onChange={handleChange}
                />
            </FormGroup>
            <FormGroup>
                <Label for="email">Email</Label>
                <Input
                    type="email"
                    name="email"
                    placeholder="Email"
                    value={loginFormData.email}
                    onChange={handleChange}
                />
            </FormGroup>
        </>
    );

    return (
        <Container>
            <Button
                className={`$activeView === "login" ? "active" : ""`}
                onClick={setLoginView}
            >
                Login
            </Button>
            <Button
                className={`$activeView === "register" ? "active" : ""`}
                onClick={setRegisterView}
            >
                Sign Up
            </Button>
            <Form onSubmit={handleLogin}>
                <FormGroup>
                    <Label for="username">Username</Label>
                    <Input
                        type="text"
                        name="username"
                        placeholder="Username"
                        value={loginFormData.username}
                        onChange={handleChange}
                    />
                </FormGroup>
                <FormGroup>
                    <Label for="username">Password</Label>
                    <Input
                        type="password"
                        name="password"
                        placeholder="Password"
                        value={loginFormData.password}
                        onChange={handleChange}
                    />
                </FormGroup>

                {activeView === "register" ? registerFields : ""}
                <Button>Submit</Button>
            </Form>
        </Container>
    );
}

export default Login;
