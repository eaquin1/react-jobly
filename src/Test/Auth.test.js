import React from 'react';
import { render } from '@testing-library/react';
import Login from '../Components/Auth/Login';
import { MemoryRouter } from 'react-router';



describe('Testing the AuthForm', () => {
  it('it renders AuthForm without crashing', () => {
    render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});