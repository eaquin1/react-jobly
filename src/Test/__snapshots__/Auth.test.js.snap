// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the AuthForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="btn-group"
    >
      <button
        class="btn btn-secondary active "
      >
        Login
      </button>
      <button
        class="btn btn-secondary  "
      >
        Sign up
      </button>
    </div>
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="username"
        >
          Username
        </label>
        <input
          class="form-control"
          name="username"
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="username"
        >
          Password
        </label>
        <input
          class="form-control"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-secondary"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
